### PR TITLE
SF Local Cluster Support for Linux VM (wsl2) when systemd is not running as PID1

### PR DESCRIPTION
**Pull request by sunil-indoria**
This features enables support for setting up SF Local cluster inside WSL2 based Ubuntu VM. Inside Ubuntu VM systemd is not running as PID1, this script change will enable that via system genie (by creating a separate namespace). Once genie is setup, installation of SF SDK happens inside the genie namespace and everything needs to be done inside genie namespace with respect to SF cluster and its management.

### CODE REVIEW

**Review by roshanroutray — APPROVED** [review]

File: scripts/SetupServiceFabric/SetupServiceFabric.sh [hunk 1]
```diff
@@ -26,6 +31,122 @@ if [[ "xenial" != "$Distribution" && "bionic" != "$Distribution" ]]; then
...
+    apt-get update; \
+    apt-get install -y apt-transport-https && \
+    apt-get update && \
+    apt-get install -y aspnetcore-runtime-5.0
+
+    echo "Adding the wsl-translinux repository"
+    apt install apt-transport-https
+
+    wget -O /etc/apt/trusted.gpg.d/wsl-transdebian.gpg https://arkane-systems.github.io/wsl-transdebian/apt/wsl-transdebian.gpg
+    chmod a+r /etc/apt/trusted.gpg.d/wsl-transdebian.gpg
```
Comment (anantshankar17) [inline scripts/SetupServiceFabric/SetupServiceFabric.sh:79]: can you avoid duplication of these steps from what you're doing and what is already happening ?

File: scripts/SetupServiceFabric/SetupServiceFabric.sh [hunk 2]
```diff
@@ -26,6 +31,122 @@ if [[ "xenial" != "$Distribution" && "bionic" != "$Distribution" ]]; then
...
+    *)
+    echo Error: Unknown option passed in: $i
+    exit 1
+    ;;
+esac
+done
+
+# Check if systemd is running as PID1, if not it should be enabled via system-genie
+pidone=$(ps --no-headers -o comm 1)
+if [ "systemd" != "$pidone" ]; then
```
Comment (anantshankar17) [inline scripts/SetupServiceFabric/SetupServiceFabric.sh:55]: what will happen when this script is not running on WSL2 but on a real Linux system ?

File: scripts/SetupServiceFabric/SetupServiceFabric.sh [hunk 3]
```diff
@@ -5,6 +5,11 @@
 # It also sets up Azure Service Fabric CLI
 #
 # Usage: sudo ./SetupServiceFabric.sh
+# Setting up Service Fabric from local .deb packages is supported for WSL2 based Linux VM
```
Comment (anantshankar17) [inline scripts/SetupServiceFabric/SetupServiceFabric.sh:8]: You should rephrase this to say the runtime and sdk are required when this installation is supposed to happen from those. This can be either of the case: WSL or real Linux system.

If this is a WSL installation or real linux system, only that check should drive if you do genie related things. Rest of the things should remain as is. please avoid copy pasting same things again and avoid duplication.

**Review by anantshankar17 — CHANGES_REQUESTED** [review]

File: scripts/SetupServiceFabric/SetupServiceFabric.sh (hunk 3 above)
Comment (sunil-indoria) [reply scripts/SetupServiceFabric/SetupServiceFabric.sh]: Additional code of block is introduced for - 
1. Installation of systemd-genie when systemd is not running as PID1, this block will setup a genie namespace.
2. Installation of sf sdk from local packages if path is provided in the script, it is supported in both WSL2 distribution and standard linux VM

For the rest of the code, re-usability has been ensured, however to ensure this genieCommand prefix is introduced which will be "genie -c" to fire the installation inside genie namespace, for standard VM is will be an empty string making the command equivalent to previous commands.

File: scripts/SetupServiceFabric/SetupServiceFabric.sh (hunk 2 above)
Comment (sunil-indoria) [reply scripts/SetupServiceFabric/SetupServiceFabric.sh]: When this script is running on a real linux system, it will not enter into genie code block and installation will happen as it was happening earlier.

File: scripts/SetupServiceFabric/SetupServiceFabric.sh (hunk 1 above)
Comment (sunil-indoria) [reply scripts/SetupServiceFabric/SetupServiceFabric.sh:79]: Done.

File: scripts/SetupServiceFabric/SetupServiceFabric.sh [hunk 4]
```diff
@@ -5,6 +5,11 @@
 # It also sets up Azure Service Fabric CLI
 #
 # Usage: sudo ./SetupServiceFabric.sh
+# Setting up Service Fabric from local service fabric runtime and sdk debain packages is all supported, both two debain packages will be needed for installation.
```
Comment (anantshankar17) [inline scripts/SetupServiceFabric/SetupServiceFabric.sh:8]: "sdk debain packages is all supported" - Fix

File: scripts/SetupServiceFabric/SetupServiceFabric.sh [hunk 5]
```diff
@@ -5,6 +5,11 @@
...
+# This script should be called with path of sf runtime and sf sdk. Below is the sample
+# sudo ./setup.sh --servicefabricruntime=/mnt/c/Users/sindoria/Downloads/servicefabric_8.2.142.2.deb --servicefabricsdk=/mnt/c/Users/sindoria/Downloads/servicefabric_sdkcommon_1.4.2.deb
```
Comment (anantshankar17) [inline scripts/SetupServiceFabric/SetupServiceFabric.sh:10]: This script should be called with path of sf runtime and sf sdk. Below is the sample
# sudo ./setup.sh --servicefabricruntime="

This script is SetupServiceFabric.sh, why do you say setup.sh ?

File: scripts/SetupServiceFabric/SetupServiceFabric.sh [hunk 6]
```diff
@@ -26,6 +31,106 @@ if [[ "xenial" != "$Distribution" && "bionic" != "$Distribution" ]]; then
...
+    echo "Genie has been started"
+fi
+
+#
+# If setup is being done inside WLS2 Distribution, then below helps in doing installation inside genie namespace if needed
+#
+genieCommand=''
+isGenieInstalled=$(apt list --installed | grep systemd-genie)
+
+if [[ ! -z "$isGenieInstalled" ]]; then
```
Comment (anantshankar17) [inline scripts/SetupServiceFabric/SetupServiceFabric.sh:102]: with all these updates, make sure the normal flow when there is no genie involved, doesnt break.

**Review by anantshankar17 — CHANGES_REQUESTED** [review]

File: scripts/SetupServiceFabric/SetupServiceFabric.sh (hunk 4 above)
Comment (sunil-indoria) [reply scripts/SetupServiceFabric/SetupServiceFabric.sh]: Fixed.

File: scripts/SetupServiceFabric/SetupServiceFabric.sh (hunk 5 above)
Comment (sunil-indoria) [reply scripts/SetupServiceFabric/SetupServiceFabric.sh]: Fixed typo

File: scripts/SetupServiceFabric/SetupServiceFabric.sh (hunk 6 above)
Comment (sunil-indoria) [reply scripts/SetupServiceFabric/SetupServiceFabric.sh:103]: will do end to end testing on VM and Linux Machine.

File: scripts/SetupServiceFabric/SetupServiceFabric.sh [hunk 7]
```diff
@@ -3,8 +3,14 @@
...
 # This script installs and sets up the Service Fabric Runtime and Common SDK.
 # It also sets up Azure Service Fabric CLI
-#
+
 # Usage: sudo ./SetupServiceFabric.sh
+# Above fetches Service Fabric Runtime and SDK along with required packages/dependencies from repositories and does the setup.
+# This script also supports installation of SDK from local .deb packages of Servie Fabric runtime and sdk. Both packages will be needed for installation and paths should be provided as parameters.
+# Below is the sample:
+# sudo ./SetupServiceFabric.sh --servicefabricruntime=/mnt/c/Users/sindoria/Downloads/servicefabric_8.2.142.2.deb --servicefabricsdk=/mnt/c/Users/sindoria/Downloads/servicefabric_sdkcommon_1.4.2.deb
+# In above scenario sf runtime is located at C:\Users\sindoria\Downloads\servicefabric_8.2.142.2.deb in windows host but in VM it will look like /mnt/c/Users/sindoria/Downloads/servicefabric_8.2.142.2.deb
```
Comment (anantshankar17) [inline scripts/SetupServiceFabric/SetupServiceFabric.sh:12]: remove all this, and just say the files should be present at that path.

File: scripts/SetupServiceFabric/SetupServiceFabric.sh [hunk 8]
```diff
@@ -26,6 +32,105 @@ if [[ "xenial" != "$Distribution" && "bionic" != "$Distribution" ]]; then
...
+if [[ ! -z "$isGenieInstalled" ]]; then
+    isGenieRunning=$(genie -r)
+    isOutsideGenie=$(genie -b)
+
+    if [[ "$isGenieRunning"=="runnning" && "$isOutsideGenie"=="outside" ]]; then
+        genieCommand="genie -c"
+    fi
+
+    # if genie is used for cluster management, current user should get permission to control service without sudo password.
+    # This enables cluster management from windows host via Powershell or LocalClusterManager.
```
Comment (anantshankar17) [inline scripts/SetupServiceFabric/SetupServiceFabric.sh:112]: these things are required only if this is running inside genie context. Fix the checks above.

**Review by anantshankar17 — CHANGES_REQUESTED** [review]

File: scripts/SetupServiceFabric/SetupServiceFabric.sh (hunk 7 above)
Comment (sunil-indoria) [reply scripts/SetupServiceFabric/SetupServiceFabric.sh]: updated the comment.

File: scripts/SetupServiceFabric/SetupServiceFabric.sh (hunk 8 above)
Comment (sunil-indoria) [reply scripts/SetupServiceFabric/SetupServiceFabric.sh]: if genie is running while setting up, then only update sudo privilege for the user.

**Review by anantshankar17 — APPROVED** [review]
Please validate all scenarios before you complete this PR: linux cluster setup on:
1. linux machine/VM
2. linux machine/VM using private runtime+SDK
3. through genie using private SDK/runtime

**Comment by sunil-indoria** [comment]
1. linux machine/VM - Installation, Cluster Setup, Cleanup, start, stop and connection work perfectly fine.
2. linux machine/VM using private runtime+SDK - while providing private bits, it creates an dependency issue (Depends: nodejs-legacy but it is not installable; Depends: libcurl3 (>= 7.16.2) but it is not going to be installed) and then pulls from repo. looking into this.
3. WSL VM - If systemd is not running, script sets up genie, Installation, Cluster Setup, Cleanup, start, stop and connection work perfectly fine.
4. WSL VM with genie using private SDK/runtime -  while providing private bits, it creates an dependency issue (Depends: nodejs-legacy but it is not installable; Depends: libcurl3 (>= 7.16.2) but it is not going to be installed) and then pulls from repo. looking into this.
